Source/WebGPU/WebGPU/CommandEncoder.mm

```diff
@@ -947,44 +949,64 @@ static bool refersToAllAspects(WGPUTextureFormat format, WGPUTextureAspect aspec
     if (rowsPerImage == WGPU_COPY_STRIDE_UNDEFINED)
         rowsPerImage = heightForMetal ?: 1;
 
-    NSUInteger sourceBytesPerImage = rowsPerImage * sourceBytesPerRow;
+    auto checkedSourceBytesPerImage = checkedProduct<NSUInteger>(rowsPerImage, sourceBytesPerRow);
+    if (checkedSourceBytesPerImage.hasOverflowed())
+        return;
+    NSUInteger sourceBytesPerImage = checkedSourceBytesPerImage.value();
 
     id<MTLTexture> mtlDestinationTexture = destinationTexture.texture();
 
     auto textureDimension = destinationTexture.dimension();
     uint32_t sliceCount = textureDimension == WGPUTextureDimension_3D ? 1 : copySize.depthOrArrayLayers;
     for (uint32_t layer = 0; layer < sliceCount; ++layer) {
-        NSUInteger destinationSlice = destinationTexture.dimension() == WGPUTextureDimension_3D ? 0 : (destination.origin.z + layer);
+        auto originPlusLayer = checkedSum<NSUInteger>(destination.origin.z, layer);
+        if (originPlusLayer.hasOverflowed())
+            return;
+        NSUInteger destinationSlice = destinationTexture.dimension() == WGPUTextureDimension_3D ? 0 : originPlusLayer.value();
         RELEASE_ASSERT(mtlDestinationTexture.parentTexture == nil);
         if (Queue::writeWillCompletelyClear(textureDimension, widthForMetal, logicalSize.width, heightForMetal, logicalSize.height, depthForMetal, logicalSize.depthOrArrayLayers))
             destinationTexture.setPreviouslyCleared(destination.mipLevel, destinationSlice);
         else
             clearTextureIfNeeded(destination, destinationSlice);
     }
 
-    NSUInteger maxSourceBytesPerRow = textureDimension == WGPUTextureDimension_3D ? (2048 * blockSize) : sourceBytesPerRow;
+    auto blockSizeTimes2048 = checkedProduct<NSUInteger>(2048, blockSize);
```

Comment:
missing if (...) return

Reply:
actually I will revert this, it is impossible to approach an overflow here (blockSize is from a function that returns at most 16)